src/nxt_application.c

```diff
@@ -343,30 +378,45 @@ nxt_app_start(nxt_task_t *task, void *data)
         ret = chdir(app_conf->working_directory);
 
         if (nxt_slow_path(ret != 0)) {
-            nxt_log(task, NXT_LOG_WARN, "chdir(%s) failed %E",
-                    app_conf->working_directory, nxt_errno);
+            nxt_app_error(task, &err, err_buf, "chdir(%s) failed %E",
+                app_conf->working_directory, nxt_errno);
 
-            return NXT_ERROR;
+            goto fail;
         }
     }
 
     if (nxt_slow_path(nxt_app_set_environment(app_conf->environment)
                       != NXT_OK))
     {
-        nxt_alert(task, "failed to set environment");
-        return NXT_ERROR;
+        nxt_app_error(task, &err, err_buf, "failed to set environment");
+        goto fail;
     }
 
     ret = nxt_app->init(task, data);
 
-    if (nxt_slow_path(ret != NXT_OK)) {
-        nxt_debug(task, "application init failed");
-
-    } else {
+    if (nxt_fast_path(ret == NXT_OK)) {
         nxt_debug(task, "application init done");
+        return NXT_OK;
     }
 
-    return ret;
+fail:
+    buf = nxt_buf_mem_alloc(rt->mem_pool,
+                            err.length+1, 0);
+
+    if (nxt_fast_path(buf != NULL)) {
+        buf->mem.free = nxt_cpymem(buf->mem.free, err.start,
+                                   err.length);
+        *buf->mem.free++ = '\0';
```

Comment:
Why is this needed? In the handler function you do the opposite: `nxt_buf_mem_used_size(&msg->buf->mem) - 1;`.

Reply:
Because I send a NULL terminated string in the message buffer.

Here: https://github.com/nginx/unit/blob/870e8e0f867603c552f10b9cb8f6941c1cf59f3e/src/nxt_controller.c#L1507
I convert it back to an `nxt_str_t` for detail. I thought that nxt_str_t doesn't need the NULL terminator. Is that correct?

Reply:
I can also just send the string bytes without NUL, as the buffer already carry the size. Is it better?

Reply:
> maybe I can just send the string bytes without NUL, as the buffer already carry the size. If this was what you meant, I think you are right, much simpler.

Yeap.